Bfiller/search endpoint

bill-filler opened the pull request:
New endpoint for searching. Usage as follows:

    Usage:
        /api/v1/search/<journal_id>/?query=<query_string>&operator=<operator>&type=<type>
    Args:
        <journal_id>: The id for Journal object to search in. If omitted will search
        all of the published Journals the requested user has access to on the given Site
        <query_string>: The text to search, should be url encoded
        <operator>: 'or' - to match any terms in the query_string (default)
                    'and' - exact match for all the terms in the query_string
        <type>: 'all' - search across all content types (default)
                'images' - search for images only
                'documents' - search in documents only
                'videos' - search in videos only
    Returns:
        List of SearchResuls objects (see SearchResultsSerializer) sorted by hit score

Known issues:

1. No pagination support: https://openedx.atlassian.net/browse/WL-1719

1. Use page titles for breadcrumbs rather than slugs: https://openedx.atlassian.net/browse/WL-1720

1. Tests needed: https://openedx.atlassian.net/browse/WL-1722

1. Add caching support: https://openedx.atlassian.net/browse/WL-1721

Also there are changes in here to subclass Image (now JournalImage) that were needed to properly index and search images. Also some changes to get_journal_about_page() stuff that wasn't working properly. I can split these out if desired.

tuchfarber commented on journals/apps/api/v1/search/views.py:
```diff
@@ -0,0 +1,198 @@
+                    self._add_to_hit_list(hit_list, page_search_results, journal_page=True)
+                    search_meta.text_count = page_search_results.count()
+
+                if search_filter == 'all' or search_filter == 'documents':
```
The document, video, and image searches seem very similar. I'd be nice if you could abstract them. Something like:

`filetype_search(queryset, model)` and have it return the search_results.

bill-filler replied:
good idea

bill-filler replied:
done

tuchfarber commented on journals/apps/api/v1/search/views.py:
```diff
@@ -0,0 +1,198 @@
+            #  search through each journal individual so we can check user permissions
+            for about_page in about_pages:
+
+                #  skip if the user doesn't have permissions for the specific journal
```
This exists in the PR I just merged, so if you want to use it, you can rebase on master

bill-filler replied:
will do

bill-filler replied:
fixed

bill-filler commented on journals/apps/api/v1/search/models.py:
```diff
@@ -0,0 +1,105 @@
+"""
+Class to store SearchResults
```
adding more clear comments to explain that it's not a DB model

bill-filler replied:
fixed

bill-filler commented on journals/apps/journals/models.py:
```diff
@@ -246,6 +248,32 @@ def data(self):
+    )
+
+
+class JournalImageRendition(AbstractRendition):
```
Seems you have to override it when you override AbstractImage, even if you are not changing it. Otherwise you get a db error where it's looking for Image instead of our subclass.
http://docs.wagtail.io/en/v1.13.2/advanced_topics/images/custom_image_model.html

tuchfarber replied:
Got it. Good to know.

bill-filler commented on journals/apps/journals/models.py:
```diff
@@ -246,6 +248,32 @@ def data(self):
         return contents
 
 
+class JournalImage(AbstractImage):
```
I can try that. Above link described overriding AbstractImage and we are doing the same thing for Document - that is, JournalDocument(AbstractDocument) not Document

tuchfarber replied:
I didn't see those docs before commenting. This is fine as is then.

bill-filler commented on journals/apps/journals/models.py:
```diff
@@ -20,14 +20,16 @@
 from wagtail.wagtailcore.fields import RichTextField, StreamField
 from wagtail.wagtailcore.models import Collection, CollectionMember, Page
 from wagtail.wagtailcore.permission_policies.collections import CollectionOwnershipPermissionPolicy
+
```
fixing

bill-filler replied:
fixed

bill-filler commented on journals/apps/journals/blocks.py:
```diff
@@ -46,27 +47,21 @@ def value_for_form(self, value):
             block=super(PDFBlock, self).render(value, context)
         )
 
-    def get_searchable_content(self, value):
```
Have a bit of a dilemma here. Anything defined in get_searchable_content() will show up as a search hit for the JournalPage which is not really what we want, we'd really want it to show up as hit for specific object type (i.e. Image, Document, Video). For example, JournalDocument has a member 'title' (value='foo') and the PDFBlock has a member 'title' (value='foo') and we search for 'foo' we'll get 2 hits which is not really desirable.

I think the behavior we want would be:
- if a model field is overriden in the block, we should only display the hit that is in the block, but have it show as a hit for the model (ie. Image, Document, Video).
Not sure how to accomplish this though.

Maybe for now I will put back the get_searchable_content() and the overriden field definition and live with the fact we will get multiple hits and hits found in the JournalPage are not linkable.

thoughts?

tuchfarber replied:
I'd prefer a match to show up as both the document/image/video and the page it resides on. As for multiple hits on the same page, could we just run results through a filter that removes duplicate pages. I'm not sure what kind of work that translates into.

bill-filler replied:
re: 'I'd prefer a match to show up as both the document/image/video and the page it resides on. '
In the current design see (https://app.zeplin.io/project/5b21820665ca1be0451f05a9/screen/5b27e5e400954cdd12ac9306)
the search results show an icon indicating the type of hit, but clicking the Title link in the result will always bring you to the page it resides on.
So for our duplicate case, you'd see two results
1) <PageIcon> <a href='page1'>My Page Title</a>
highlight: 'My Document Title'
2) <DocumentIcon> <a href='page1'>My Document Ttile</a>
highlight: 'My Document Title'

I would assume we'd only want to see 2)

bill-filler commented on journals/apps/api/v1/search/views.py:
```diff
@@ -0,0 +1,198 @@
+)
+from journals.apps.api.serializers import SearchResultsSerializer
+
+from .models import SearchResults, SearchHit, SearchMetaData
```
will fix

bill-filler replied:
fixed

bill-filler commented on journals/apps/api/v1/search/views.py:
```diff
@@ -0,0 +1,198 @@
+
+                #  skip if the user doesn't have permissions for the specific journal
+                if not JournalAccess.user_has_access(request.user, about_page.journal):
+                    logger.warning(
```
will change to debug

bill-filler replied:
this is now gone as we check access up front and only loop through journals user has access to, not all of them

bill-filler commented on journals/apps/api/v1/search/views.py:
```diff
@@ -0,0 +1,198 @@
+                #  Only include pages beneath our root page that are live and public
+                base_page_query = JournalPage.objects.live().public().descendant_of(about_page)
+
+                if search_filter == 'all':
```
will fix

bill-filler replied:
done

bill-filler commented on journals/apps/api/v1/search/views.py:
```diff
@@ -0,0 +1,198 @@
+                base_page_query = JournalPage.objects.live().public().descendant_of(about_page)
+
+                if search_filter == 'all':
+                    # Search pages which will yeild hits for text/HTML/RawHTML
```
will fix

bill-filler replied:
done

bill-filler commented:
@tuchfarber 
ready for another review. I added back in the get_searchable_content() in the block and added fields so the blocks could override value specified in the model.

tuchfarber commented on journals/apps/api/v1/search/views.py:
```diff
@@ -0,0 +1,247 @@
+        about_pages = self._get_journals_for_user(request, journal_id)
+
+        if not about_pages:
+            return None
```
return either `[]` or `about_pages` so we can pass an empty list back.

tuchfarber commented on journals/apps/api/v1/search/views.py:
```diff
@@ -0,0 +1,247 @@
+        # return the list iterator in descending order with highest hit score first
+        return SearchResults(search_meta, reversed(hit_list))
+
+    def _filetype_search(self, base_page_query, query_string, search_operator, type_class):
```
This is great!

bill-filler commented:
@tuchfarber 
comments addressed, merging..
